File: mijoz_app/lib/models/catalog.dart
class Product {
  final int id;
  final int restaurantId;
  final int categoryId;
  final String nameUz;
  final String nameRu;
  final String? descriptionUz;
  final String? descriptionRu;
  final String? imageUrl;
  final int price;
  final String? unit;
  final bool isAvailable;

  Product({
    required this.id, required this.restaurantId, required this.categoryId,
    required this.nameUz, required this.nameRu, this.descriptionUz,
    this.descriptionRu, this.imageUrl, required this.price,
    this.unit, required this.isAvailable,
  });

  factory Product.fromJson(Map<String, dynamic> json) => Product(
    id: json['id'], restaurantId: json['restaurant_id'], categoryId: json['category_id'],
    nameUz: json['name_uz'], nameRu: json['name_ru'], descriptionUz: json['description_uz'],
    descriptionRu: json['description_ru'], imageUrl: json['image_url'],
    price: json['price'], unit: json['unit'], isAvailable: json['is_available'] ?? true,
  );
}

class Subcategory {
  final int id;
  final String nameUz;
  final String nameRu;
  final String? imageUrl;
  final int sortOrder;
  final List<Product> products;

  Subcategory({
    required this.id, required this.nameUz, required this.nameRu,
    this.imageUrl, required this.sortOrder, required this.products,
  });

  factory Subcategory.fromJson(Map<String, dynamic> json) => Subcategory(
    id: json['id'], nameUz: json['name_uz'], nameRu: json['name_ru'],
    imageUrl: json['image_url'], sortOrder: json['sort_order'],
    products: (json['products'] as List?)?.map((e) => Product.fromJson(e)).toList() ?? [],
  );
}

class Category {
  final int id;
  final int? groupId;
  final String nameUz;
  final String nameRu;
  final String? imageUrl;
  final int sortOrder;
  final List<Subcategory> subcategories;

  Category({
    required this.id, this.groupId, required this.nameUz, required this.nameRu,
    this.imageUrl, required this.sortOrder, required this.subcategories,
  });

  factory Category.fromJson(Map<String, dynamic> json) => Category(
    id: json['id'], groupId: json['group_id'], nameUz: json['name_uz'],
    nameRu: json['name_ru'], imageUrl: json['image_url'], sortOrder: json['sort_order'],
    subcategories: (json['subcategories'] as List?)?.map((e) => Subcategory.fromJson(e)).toList() ?? [],
  );
}

class CategoryGroup {
  final int id;
  final String nameUz;
  final String nameRu;
  final int sortOrder;

  CategoryGroup({required this.id, required this.nameUz, required this.nameRu, required this.sortOrder});

  factory CategoryGroup.fromJson(Map<String, dynamic> json) => CategoryGroup(
    id: json['id'], nameUz: json['name_uz'], nameRu: json['name_ru'], sortOrder: json['sort_order'],
  );
}

class RestaurantDetail {
  final int id;
  final String name;
  final List<Category> categories;
  final List<CategoryGroup> categoryGroups;
  final int deliveryFee;
  final int minOrder;
  final int avgDeliveryMinutes;

  RestaurantDetail({
    required this.id, required this.name, required this.categories,
    required this.categoryGroups, required this.deliveryFee,
    required this.minOrder, required this.avgDeliveryMinutes,
  });

  factory RestaurantDetail.fromJson(Map<String, dynamic> json) => RestaurantDetail(
    id: json['id'], name: json['name'],
    categories: (json['categories'] as List?)?.map((e) => Category.fromJson(e)).toList() ?? [],
    categoryGroups: (json['category_groups'] as List?)?.map((e) => CategoryGroup.fromJson(e)).toList() ?? [],
    deliveryFee: json['delivery_fee'] ?? 0,
    minOrder: json['min_order'] ?? 0,
    avgDeliveryMinutes: json['avg_delivery_minutes'] ?? 0,
  );
}

File: backend/app/schemas/catalog.py
from pydantic import BaseModel


class ProductOut(BaseModel):
    id: int
    restaurant_id: int
    category_id: int
    name_uz: str
    name_ru: str
    description_uz: str | None = None
    description_ru: str | None = None
    image_url: str | None = None
    price: int
    cost: int = 0
    stock: float = 0
    unit: str = "dona"
    low_stock_threshold: float = 10
    is_available: bool

    class Config:
        from_attributes = True


class CategoryOut(BaseModel):
    id: int
    parent_id: int | None = None
    # Faqat top-level kategoriyalarda — qaysi Title (CategoryGroup) ostida ko'rsatilishi.
    group_id: int | None = None
    name_uz: str
    name_ru: str
    image_url: str | None = None
    sort_order: int

    class Config:
        from_attributes = True


class CategoryGroupOut(BaseModel):
    """Title — bosh sahifada bir nechta kategoriyani guruhlaydigan sarlavha."""
    id: int
    name_uz: str
    name_ru: str
    sort_order: int

    class Config:
        from_attributes = True


class SubcategoryOut(CategoryOut):
    products: list[ProductOut] = []


class CategoryWithSubcategories(CategoryOut):
    subcategories: list[SubcategoryOut] = []


class RestaurantOut(BaseModel):
    id: int
    name: str
    description_uz: str | None = None
    description_ru: str | None = None
    logo_url: str | None = None
    cover_url: str | None = None
    address: str | None = None
    owner_name: str | None = None
    phones: list[str] = []
    socials: dict[str, str] = {}
    lat: float | None = None
    lng: float | None = None
    is_active: bool
    is_open: bool
    rating: float
    delivery_fee: int
    min_order: int
    avg_delivery_minutes: int

    class Config:
        from_attributes = True


class StoreSettingsIn(BaseModel):
    """Do'kon identifikatsiya sozlamalari (admin sozlamalar sahifasi)."""
    name: str
    description_uz: str | None = None
    description_ru: str | None = None
    logo_url: str | None = None
    cover_url: str | None = None
    address: str | None = None
    owner_name: str | None = None
    phones: list[str] = []
    socials: dict[str, str] = {}
    lat: float | None = None
    lng: float | None = None
    # Yetkazish: min_order = bepul chegara (so'm); delivery_fee = so'm/km.
    min_order: int = 50_000
    delivery_fee: int = 2_000


class RestaurantDetail(RestaurantOut):
    categories: list[CategoryWithSubcategories] = []
    # Title'lar — mijoz ilovasi shu ro'yxat + har categoriyaning group_id'si orqali
    # kategoriyalarni sarlavha ostida guruhlaydi (group_id=None — sarlavhasiz).
    category_groups: list[CategoryGroupOut] = []


# ── Admin write schemas ──────────────────────────────────────────
class RestaurantIn(BaseModel):
    name: str
    description_uz: str | None = None
    description_ru: str | None = None
    logo_url: str | None = None
    cover_url: str | None = None
    is_active: bool = True
    is_open: bool = True
    delivery_fee: int = 2000   # so'm/km
    min_order: int = 50_000    # bepul yetkazish chegarasi
    avg_delivery_minutes: int = 40


class CategoryIn(BaseModel):
    parent_id: int | None = None
    group_id: int | None = None
    name_uz: str
    name_ru: str
    image_url: str | None = None
    sort_order: int = 0


class CategoryGroupIn(BaseModel):
    name_uz: str
    name_ru: str
    sort_order: int = 0


class ProductIn(BaseModel):
    category_id: int
    name_uz: str
    name_ru: str
    description_uz: str | None = None
    description_ru: str | None = None
    image_url: str | None = None
    price: int
    cost: int = 0
    stock: float = 0
    unit: str = "dona"
    low_stock_threshold: float = 10
    is_available: bool = True
    sort_order: int = 0
